src/sinks/util/zstd.rs

```diff
@@ -0,0 +1,67 @@
+use std::{fmt::Display, io};
+
+use super::buffer::compression::CompressionLevel;
+
+#[derive(Debug)]
+pub struct ZstdCompressionLevel(i32);
+
+impl From<CompressionLevel> for ZstdCompressionLevel {
+    fn from(value: CompressionLevel) -> Self {
+        let val: i32 = match value {
+            CompressionLevel::None => 0,
+            CompressionLevel::Default => zstd::DEFAULT_COMPRESSION_LEVEL,
+            CompressionLevel::Best => 21,
+            CompressionLevel::Fast => 1,
+            CompressionLevel::Val(v) => v as i32,
+        };
+        ZstdCompressionLevel(val)
+    }
+}
+
+impl Display for ZstdCompressionLevel {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+pub struct ZstdEncoder<W: io::Write> {
+    inner: zstd::Encoder<'static, W>,
+}
+
+impl<W: io::Write> ZstdEncoder<W> {
+    pub fn new(writer: W, level: ZstdCompressionLevel) -> io::Result<Self> {
+        let encoder = zstd::Encoder::new(writer, level.0)?;
+        Ok(Self { inner: encoder })
+    }
+
+    pub fn finish(self) -> io::Result<W> {
+        self.inner.finish()
+    }
+
+    pub fn get_ref(&self) -> &W {
+        self.inner.get_ref()
+    }
+}
+
+impl<W: io::Write> io::Write for ZstdEncoder<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.inner.write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner.flush()
+    }
+}
+
+impl<W: io::Write + std::fmt::Debug> std::fmt::Debug for ZstdEncoder<W> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ZstdEncoder")
+            .field("inner", &self.get_ref())
+            .finish()
+    }
+}
+
+/// Safety:
+/// 1. There is no sharing references to zstd encoder. `Write` requires unique reference, and `finish` moves the instance itself.
+/// 2. Sharing only internal writer, which implements `Sync`
+unsafe impl<W: io::Write + Sync> Sync for ZstdEncoder<W> {}
```

Comment:
Rust is usually able to auto-derive `Sync` when it is safe to do so. Do you know why this manual `impl` is necessary?

Reply:
It's required to be a `Sync` by `DatadogMetricsRequestBuilder`, which owns a `Compressor` instance.
`zstd::Encoder` is not `Sync` because of the internal struct `ZSTD_CCtx_s `, which does not implement `Sync`.
The details are here https://github.com/gyscos/zstd-rs/issues/186.

So, I think, it's pretty safe to declare `unsafe Sync` here, because:
 1. we are not sharing`zstd::Encoder`, only internal `impl Write`, which should implement `Sync`;
 2. using the `ZstdEncoder ` itself for compressing requires `&mut self`, which also means that we can't share it.

Same workaround is using by `async-compression` - zstd encoder is wrapped by `Unshared` struct, which gives only a mutable ref to encoder.

https://github.com/Nullus157/async-compression/blob/6d453d88bb2eb301cbbc9a60bb2b61d3a88be14d/src/codec/zstd/encoder.rs#LL7C17-L7C17

Reply:
Works for me. If we end up having more of these, it would probably be worth bringing in the `Unshared` wrapper type to consolidate the logic, but otherwise this should be fine as-is.